cds_ils/importer/views.py

```diff
@@ -0,0 +1,169 @@
+            for entry in entries:
+                if not entry.error:
+                    reports.append({
+                        "index": entry.entry_index,
+                        "success": True,
+                        "report": {
+                            "ambiguous_documents": entry.ambiguous_documents,
+                            "ambiguous_eitems": entry.ambiguous_eitems,
+                            "created_document": entry.created_document,
+                            "created_eitem": entry.created_eitem,
+                            "updated_document": entry.updated_document,
+                            "updated_eitem": entry.updated_eitem,
+                            "deleted_eitems": entry.deleted_eitems,
+                            "series": entry.series,
+                            "fuzzy_documents": entry.fuzzy_documents,
+                        }
+                    })
+                else:
+                    reports.append({
+                        "index": entry.entry_index,
+                        "success": False,
+                        "message": entry.error,
+                    })
+            obj["reports"] = reports
+            return obj
+        else:
+            raise Exception("Task not found")  # TODO
+
+    @blueprint.route('/importer', methods=['POST'])
+    @need_permissions("document-importer")
+    def importer():
+
+        @copy_current_request_context
+        def process_data(log_id, file, source_type, provider):
+            with open(app.config["IMPORTER_UPLOADS_PATH"] + "/"
+                      + file.filename) as f:
+                import_from_xml(log_id, f, source_type, provider)
+
+        if request.files:
+            file = request.files["file"]
+            provider = request.form.get("provider", None)
+            mode = request.form.get("mode", None)
+
+            if not provider:
+                flash('Missing provider')
+                raise RequestError('Missing provider', 400)
+
+            if not mode:
+                flash('Missing mode')
+                raise RequestError('Missing mode', 400)
+
+            if allowed_files(file.filename):
+                original_filename = file.filename
+                file.filename = rename_file(file.filename)
+                filename = secure_filename(file.filename)
+                file.save(os.path.join(app.config["IMPORTER_UPLOADS_PATH"],
+                                       filename))
+
+                # TODO move this to API
+                source_type = "marcxml"
+                log = ImporterTaskLog.create(dict(
+                    agent=ImporterAgent.USER,
+                    provider=provider,
+                    source_type=source_type,
+                    mode=ImporterMode.CREATE if mode == "create"
+                    else ImporterMode.UPDATE,
+                    original_filename=original_filename,
+                ))
+
+                t = Thread(target=process_data,
```

Comment:
why a new thread in a request context? What is the advantage?

Reply:
It's used to launch the importing process and to be able to return a 200 (if everything ok) at the same time. Otherwise we would be waiting for the response